py-polars/polars/dataframe/group_by.py

```diff
@@ -311,16 +311,12 @@ def map_groups(self, function: Callable[[DataFrame], DataFrame]) -> DataFrame:
         ...     pl.int_range(pl.len()).shuffle().over("color") < 2
         ... )  # doctest: +IGNORE_RESULT
         """
-        if self.named_by:
-            msg = "cannot call `map_groups` when grouping by named expressions"
-            raise TypeError(msg)
-        if not all(isinstance(c, str) for c in self.by):
-            msg = "cannot call `map_groups` when grouping by an expression"
-            raise TypeError(msg)
-
-        return self.df.__class__._from_pydf(
-            self.df._df.group_by_map_groups(
-                list(self.by), function, self.maintain_order
+        self.by_str_cols = self.df.lazy().select(*self.by, **self.named_by).columns
+        df = self.df.with_columns(*self.by, **self.named_by)
```

Comment:
I think this might handle name clashes incorrectly, but I'm not sure. Could you add some tests for this?

Reply:
I don't think it executes twice. In line 314 it uses a lazy select so that it doesn't materialize anything except a new schema from which I take the columns as a list of column names. In line 315, it does eagerly materialize those expressions into columns before the `group_by_map_groups`.